### PR TITLE
Make SystemTime mockable

### PR DESCRIPTION
**Pull request by gretchenfrage**
Only 1 of this and #2089 should be merged, and the other should be closed. To help automate this process, I am marking them as closing each other.

Closes #2089

---

We already do this for Instants. However, there are currently two places in main that we call SystemTime::now from proto. This PR makes them mockable for tests.

- `quinn::Runtime` gains new `system_now(&self) -> SystemTime` method. This has a default implementation, so no semver breakage for quinn.
- `proto::Endpoint::handle` gains a `system_now: SystemTime` argument. This causes semver breakage for proto.
- The `SystemTime` provided in `handle` is stored in `Incoming.received_at_system`, just like `Incoming.received_at`, and may be utilized in `Endpoint::retry`.
- Tests are adapted by making each `TestEndpoint` have its own `Instant` -> `SystemTime` conversion.

### CODE REVIEW

File: quinn/src/runtime.rs [hunk 1]
```diff
@@ -25,6 +25,12 @@ pub trait Runtime: Send + Sync + Debug + 'static {
     fn now(&self) -> Instant {
         Instant::now()
     }
+    /// Look up the current [`SystemTime`]
```
Comment (Ralith) [inline quinn/src/runtime.rs:28]: style: empty line between definitions

**Review by Ralith — CHANGES_REQUESTED** [review]
Unfortunately, as `quinn-proto` is a public dependency, breaking changes to it are necessarily breaking changes to `quinn`. That said, minimizing API churn is still valuable regardless of semver.

It feels a little funny to be invoking `SystemTime::now` on every received packet receipt when it is almost never actually needed. I wonder if a trait object in `EndpointConfig` might be a more proportionate interface, and coincidentally a more semver-friendly one besides? Interested in perspectives.

**Comment by gretchenfrage** [comment]
> I wonder if a trait object in EndpointConfig might be a more proportionate interface, and coincidentally a more semver-friendly one besides?

Maybe, but it makes it more complicated to mock. At worst, mocking could resort to feeling like they have to create an implementation that accesses thread-locals to thread this through method calls in a side channel.

**Comment by gretchenfrage** [comment]
> > I wonder if a trait object in EndpointConfig might be a more proportionate interface, and coincidentally a more semver-friendly one besides?
> 
> Maybe, but it makes it more complicated to mock. At worst, mocking could resort to feeling like they have to create an implementation that accesses thread-locals to thread this through method calls in a side channel.

A secret third option would be to pass in a callback to `handle_packet` (an `impl FnOnce() -> SystemTime` or equivalent trait).

File: quinn/src/runtime.rs (hunk 1 above)
Comment (gretchenfrage) [reply quinn/src/runtime.rs:32]: Then I shall fix it for the others in this file too.

**Comment by gretchenfrage** [comment]
> > > I wonder if a trait object in EndpointConfig might be a more proportionate interface, and coincidentally a more semver-friendly one besides?
> > 
> > 
> > Maybe, but it makes it more complicated to mock. At worst, mocking could resort to feeling like they have to create an implementation that accesses thread-locals to thread this through method calls in a side channel.
> 
> A secret third option would be to pass in a callback to `handle_packet` (an `impl FnOnce() -> SystemTime` or equivalent trait).

See 8f1408e008dcefcef15fb62b0d0188d8db946017 (pushed to side branch as a demonstration of what it could be like)

**Comment by Ralith** [comment]
> At worst, mocking could resort to feeling like they have to create an implementation that accesses thread-locals to thread this through method calls in a side channel.

Surely simple shared state would be adequate?

**Comment by gretchenfrage** [comment]
Yeah we should probably just go with #2089 it's way simpler.

**Comment by Ralith** [comment]
That does look much more manageable, thanks!

**Comment by gretchenfrage** [comment]
Closing in favor of #2089

**Pull request closed without merging**